src/class.c

```diff
@@ -1634,22 +1661,27 @@ bool InterfaceDeclaration::isBaseOf(ClassDeclaration *cd, int *poffset)
     {
         BaseClass *b = cd->interfaces[j];
 
-        //printf("\tbase %s\n", b->base->toChars());
-        if (this == b->base)
+        //printf("\tX base %s\n", b->sym->toChars());
+        if (this == b->sym)
         {
             //printf("\tfound at offset %d\n", b->offset);
             if (poffset)
             {
                 *poffset = b->offset;
-                if (j && cd->isInterfaceDeclaration())
+                // The conversion is interface(cd) to interface(this).
+                if (cd->isInterfaceDeclaration())
```

Comment:
> though I believe that C++ doesn't have to do it with virtual inheritence. (Not sure these are exactly the same things, though.)

Is that right? I'm not really sure about the C++ class layout, but I can think a diamond class hierarchy with virtual inheritance would need dynamic cast.

> the vtbl for M completely contains the interface for I, so a dynamic cast should be unnecessary here.

I think D's interface inheritance is always equivalent with virtual inheritance in C++. A target base interface type might be a top of diamond in the actual implemented class hierarchy. Therefore compiler cannot optimize such the upcast by using a static cast.

I think it's current limitation comes from the ABI spec. If compiler emits more detailed class hierarchy information into RTTI structures, such the cast cost might be reduced. But it would be a ABI enhancement.